libfwupdplugin/fu-bios-attrs.c

```diff
@@ -0,0 +1,334 @@
+/*
+ * Copyright (C) 2022 Mario Limonciello <mario.limonciello@amd.com>
+ *
+ * SPDX-License-Identifier: LGPL-2.1+
+ */
+
+#define G_LOG_DOMAIN "FuBiosAttrs"
+
+#include "config.h"
+
+#include "fu-bios-attrs-private.h"
+#include "fu-path.h"
+#include "fu-string.h"
+
+struct _FuBiosAttrs {
+	GObject parent_instance;
+	gboolean kernel_bug_shown;
+	GPtrArray *attrs;
+};
+
+G_DEFINE_TYPE(FuBiosAttrs, fu_bios_attrs, G_TYPE_OBJECT)
+
+static void
+fu_bios_attrs_finalize(GObject *obj)
+{
+	FuBiosAttrs *self = FU_BIOS_ATTRS(obj);
+	g_ptr_array_unref(self->attrs);
+	G_OBJECT_CLASS(fu_bios_attrs_parent_class)->finalize(obj);
+}
+
+static void
+fu_bios_attrs_class_init(FuBiosAttrsClass *klass)
+{
+	GObjectClass *object_class = G_OBJECT_CLASS(klass);
+	object_class->finalize = fu_bios_attrs_finalize;
+}
+
+static gboolean
+fu_bios_attr_get_key(FwupdBiosAttr *attr, const gchar *key, gchar *value_out, GError **error)
+{
+	g_autofree gchar *tmp = NULL;
+
+	g_return_val_if_fail(FWUPD_IS_BIOS_ATTR(attr), FALSE);
+	g_return_val_if_fail(key != NULL, FALSE);
+	g_return_val_if_fail(&value_out != NULL, FALSE);
+
+	tmp = g_build_filename(fwupd_bios_attr_get_path(attr),
+			       fwupd_bios_attr_get_name(attr),
+			       key,
+			       NULL);
+	if (!g_file_get_contents(tmp, &value_out, NULL, error))
+		return FALSE;
+	g_strchomp(value_out);
+	return TRUE;
+}
+
+static gboolean
+fu_bios_attr_set_description(FwupdBiosAttr *attr)
+{
+	g_autofree gchar *data = NULL;
+	g_autoptr(GError) error_local = NULL;
+
+	g_return_val_if_fail(FWUPD_IS_BIOS_ATTR(attr), FALSE);
+
+	if (g_strcmp0(fwupd_bios_attr_get_name(attr), FWUPD_BIOS_ATTR_PENDING_REBOOT) == 0) {
+		fwupd_bios_attr_set_description(attr, "Settings will apply after system reboots");
+		return TRUE;
+	}
+	if (!fu_bios_attr_get_key(attr, "display_name", data, &error_local)) {
+		g_debug("%s", error_local->message);
+		return FALSE;
+	}
+	fwupd_bios_attr_set_description(attr, data);
+
+	return TRUE;
+}
+
+static void
+fu_bios_attr_set_integer_attrs(FwupdBiosAttr *attr)
+{
+	g_autofree gchar *lower = NULL, *upper = NULL, *scalar = NULL;
```

Comment:
I don't think does what you think it does; you need g_autofree for each.

Reply:
fixed